File: data/download_data.py
import yfinance as yf
import pandas as pd
import os
import argparse

def download_data(tickers, start_date, end_date, output_dir):

    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    print(f"Downloading data for: {tickers}")
    data = yf.download(tickers, start=start_date, end=end_date, group_by='ticker')
    
    if len(tickers) == 1:
        ticker = tickers[0]
        file_path = os.path.join(output_dir, f"{ticker}.csv")
        data.to_csv(file_path)
        print(f"Saved {ticker} to {file_path}")
    else:
        for ticker in tickers:
            ticker_data = data[ticker]
            file_path = os.path.join(output_dir, f"{ticker}.csv")
            ticker_data.to_csv(file_path)
            print(f"Saved {ticker} to {file_path}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Download financial data.')
    parser.add_argument('--tickers', nargs='+', default=['SPY', 'AAPL', 'BTC-USD'], help='List of tickers')
    parser.add_argument('--start', type=str, default='2020-01-01', help='Start date (YYYY-MM-DD)')
    parser.add_argument('--end', type=str, default='2023-01-01', help='End date (YYYY-MM-DD)')
    parser.add_argument('--output', type=str, default='data/raw', help='Output directory')
    
    args = parser.parse_args()
    
    download_data(args.tickers, args.start, args.end, args.output)

File: data/download_data.R
if (!require("quantmod")) install.packages("quantmod")
if (!require("tidyverse")) install.packages("tidyverse")

library(quantmod)
library(tidyverse)

download_data_r <- function(tickers, start_date, end_date, output_dir) {
  
  if (!dir.exists(output_dir)) {
    dir.create(output_dir, recursive = TRUE)
  }
  
  for (ticker in tickers) {
    print(paste("Downloading", ticker))
    tryCatch({
      getSymbols(ticker, src = "yahoo", from = start_date, to = end_date, auto.assign = TRUE)
      data <- get(ticker)
      
      df <- data.frame(Date = index(data), coredata(data))
      
      file_path <- file.path(output_dir, paste0(ticker, "_R.csv"))
      write.csv(df, file_path, row.names = FALSE)
      print(paste("Saved", ticker, "to", file_path))
      
    }, error = function(e) {
      print(paste("Error downloading", ticker, ":", e$message))
    })
  }
}

tickers <- c("SPY", "AAPL", "BTC-USD")
start_date <- "2020-01-01"
end_date <- "2023-01-01"
output_dir <- "data/raw"

download_data_r(tickers, start_date, end_date, output_dir)
